thread-rcu/rcu.h

```diff
@@ -0,0 +1,214 @@
+#ifndef __RCU_H__
+#define __RCU_H__
+
+/* lock primitives from pthread and compiler primitives */
+
+#include <pthread.h>
+
+typedef pthread_mutex_t spinlock_t;
+
+#define SPINLOCK_INIT PTHREAD_MUTEX_INITIALIZER
+static inline void spin_lock(spinlock_t *sp)
+{
+    int ret;
+
+    ret = pthread_mutex_lock(sp);
+    if (ret != 0) {
+        fprintf(stderr, "spin_lock:pthread_mutex_lock %d\n", ret);
+        abort();
+    }
+}
+
+static inline void spin_unlock(spinlock_t *sp)
+{
+    int ret;
+
+    ret = pthread_mutex_unlock(sp);
+    if (ret != 0) {
+        fprintf(stderr, "spin_unlock:pthread_mutex_unlock %d\n", ret);
+        abort();
+    }
+}
+
+#define current_tid() (unsigned int) pthread_self()
+
+#define ACCESS_ONCE(x) (*(volatile __typeof__(x) *) &(x))
+#define READ_ONCE(x)                         \
+    ({                                       \
+        __typeof__(x) ___x = ACCESS_ONCE(x); \
+        ___x;                                \
+    })
+#define WRITE_ONCE(x, val)      \
+    do {                        \
+        ACCESS_ONCE(x) = (val); \
+    } while (0)
+#define barrier() __asm__ __volatile__("" : : : "memory")
+#define smp_mb() __asm__ __volatile__("mfence" : : : "memory")
```

Comment:
The `mfence` is specific to x86, and I would like to stick to C11 Atomics for portability. Think of the use of [atomic_thread_fence](https://en.cppreference.com/w/cpp/atomic/atomic_thread_fence).